quill/ecs/src/component.rs

```diff
@@ -0,0 +1,67 @@
+use std::{alloc::Layout, any::TypeId, sync::Arc};
+
+use crate::space::MemorySpace;
+
+/// Type ID of a component.
+///
+/// Supports both Rust types and arbitrary
+/// "opaque" types identified by a `u64`.
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash, PartialOrd, Ord)]
+pub struct ComponentTypeId(ComponentTypeIdInner);
+
+impl ComponentTypeId {
+    pub fn of<T: 'static>() -> Self {
+        Self(ComponentTypeIdInner::Rust(TypeId::of::<T>()))
+    }
+
+    pub fn opaque(id: u64) -> Self {
+        Self(ComponentTypeIdInner::Opaque(id))
+    }
+}
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash, PartialOrd, Ord)]
+enum ComponentTypeIdInner {
+    Rust(TypeId),
```

Comment:
Maybe just use `Opaque` and use a `FnvHashMap` to map from type id to opaque id.

Reply:
How would that be better?

Reply:
We have to chose a true id at one point, I suggest we chose `Opaque` as the true id and map `TypeId` to the true id(`Opaque`). This indirection have happen at some point.

Reply:
What do you mean by "true id" and why is it needed?

Reply:
"true id" one universal identifier for components, no distinction between local ie `Any::TypeId` or remote ids.

Reply:
And **why is it needed**?